boot/boot.go

```diff
@@ -455,31 +454,43 @@ func makeBootable20RunMode(model *asserts.Model, rootdir string, bootWith *Boota
 		return fmt.Errorf("cannot write modeenv: %v", err)
 	}
 
-	// get the ubuntu-boot bootloader
+	ubuntuBootPartition := filepath.Join(runMnt, "ubuntu-boot")
+
+	// get the ubuntu-boot grub and extract the kernel there
 	opts := &bootloader.Options{
 		// TODO:UC20: we use "recovery: true" here because on
 		// the partition the file layout of ubuntu-boot looks
 		// the same as ubuntu-seed.
 		// TODO:UC20: need a better name than recovery
 		Recovery: true,
+		// extract kernel assets to the ubuntu-boot partition
+		KernelExtractionDir: ubuntuBootPartition,
 	}
-	bl, err := bootloader.Find(filepath.Join(runMnt, "ubuntu-boot"), opts)
+	bl, err := bootloader.Find(ubuntuBootPartition, opts)
 	if err != nil {
 		return fmt.Errorf("internal error: cannot find run system bootloader: %v", err)
 	}
-	// TODO:UC20: using the UC16/18 grubenv style until we have
-	// a UC20 grub.cfg and corresponding snapd early boot
-	// code
+
+	// extract the kernel first, then mark kernel_status ready in case we get
+	// rebooted in between
+	kernelf, err := snap.Open(bootWith.KernelPath)
+	if err != nil {
+		return err
+	}
+
+	// TODO:UC20: support non uefi kernel assets
+	// for now we will only support kernel.efi
+	err = bl.ExtractKernelAssets(bootWith.Kernel, kernelf, []string{"kernel.efi"})
```

Comment:
Feels like actual assets list need not be exposed here and the `Bootloader` is responsible for knowing what to extract. Maybe we need to pass `bootloader.Options` to `ExtractKernelAssets` like we already do in `InstallBootConfig()` and check the Recovery flag there.

Reply:
Well I thought about doing it that way, but the reason I went with it this way is that we may find ourselves in the same situation for non-grub bootloaders where we have a single bootloader which extracts different assets depending on how it's used. 
Just as an example, perhaps at some point we will have a customer need to support UC20 on non-UEFI hardware, and in this case it's easier to support since we would just change the list of assets passed in. 

I can certainly refactor it if folks think that's a better way to do this.

Reply:
let's have a boringly named  ExtractedRunKernelImage option in opts here instead,  we also have still the problem that Recovery: true is abused here, I had a suggestion somewhere about that

Reply:
BootNotMounted was my original suggestion about that 2nd point,  NoSlashBoot was also work

Reply:
Okay, I reverted the change adding the list of assets in the arguments to ExtractKernelAssets and added a bool `ExtractedRunKernelImage` which is part of opts passed to `bootloader.Find` and then used by grub to determine which assets to extract. Let me know if I misinterpreted what was desired here.